qcow/src/qcow.rs

```diff
@@ -775,7 +775,8 @@ impl QcowFile {
             refblock_clusters: u64,
             pointers_per_cluster: u64,
         ) -> Result<Vec<u64>> {
-            let refcount_table_entries = div_round_up_u64(refblock_clusters, pointers_per_cluster);
+            let refcount_clusters = div_round_up_u64(refblock_clusters, pointers_per_cluster);
+            let refcount_table_entries = refcount_clusters + refblock_clusters;
```

Comment:
I'm not sure this is the right thing to do. I mean you end up having `refcount_table_entries` equal to the total amount of clusters needed to reference data clusters. Is that what we need here?
I think the `ref_table` is the list of u64 addresses pointing to `refblocks`, which is why `refcount_table_entries` should be the number of clusters of the second level.

Also, please provide a clear commit message to explain what was wrong and how you fixed it.

Reply:
Okay, thanks for the explanation, that sounds fine to me.
Please update the commit message and I'll approve the PR.

Reply:
sorry, something wrong happend, 
thread 'tests::seek_data_direct' panicked at 'slice index starts at 32768 but ends at 10'

qcow::QcowFile::rebuild_refcounts::write_refblocks() at  ./src/qcow.rs:822:33  
```
for (i, refblock_addr) in ref_table.iter().enumerate() {
                // Write a block of refcounts to the location indicated by refblock_addr.
                let refblock_start = i * (refcount_block_entries as usize);
                let refblock_end = min(
                    refcounts.len(),
                    refblock_start + refcount_block_entries as usize,
                );
                **let refblock = &refcounts[refblock_start..refblock_end];**
                raw_file
                    .write_refcount_block(*refblock_addr, refblock)
                    .map_err(Error::WritingHeader)?;

                // If this is the last (partial) cluster, pad it out to a full refblock cluster.
                if refblock.len() < refcount_block_entries as usize {
                    let refblock_padding =
                        vec![0u16; refcount_block_entries as usize - refblock.len()];
                    raw_file
                        .write_refcount_block(
                            *refblock_addr + refblock.len() as u64 * 2,
                            &refblock_padding,
                        )
                        .map_err(Error::WritingHeader)?;
                }
            }

```

Reply:
vector index refblock_start is greater than index refblock_end

Reply:
Well it seems like your patch isn't doing what you were expecting then.

FWIW, I've tried patching with the following:
```rust
let refcount_table_entries = refblock_clusters;
```
and it works fine (the unit tests are passing).

Reply:
Maybe you are right.  Here use  round_up to caculate parameter refblock_clusters, and  it should be enough to cover the few clusters used by refcount table：
`let refblock_clusters = div_round_up_u64(max_valid_cluster_index, refcount_block_entries);`